### PR TITLE
Jsdoc 3

### PR DESCRIPTION
**Pull request by mzummo**
@dtsnell4 Please Review

### CODE REVIEW

File: src/common/HTMLWidget.js [hunk 1]
```diff
@@ -37,18 +72,54 @@
         return retVal;
     };
 
+    /**
+     * Returns the true height of an element by returning css width - (minus) calcFrameHeight().
```
Comment (dtsnell4) [inline src/common/HTMLWidget.js:76]: Should say "returning css height"

File: src/common/Widget.js [hunk 2]
```diff
@@ -649,11 +964,65 @@
...
+     * @function
+     */
 
-    //  Util  ---
+    /**
+     * Limits rate at which a function can be fired. (Normally used for "resize / positioning" events)
+     * @method debounce
+     * @memberof common_Widget
+     * @instance
+     * @param {Widget~DebounceCb} [callback] Function that needs it fire rate limited.
```
Comment (dtsnell4) [inline src/common/Widget.js:1011]: "it" -> "it's"

File: src/common/Widget.js [hunk 3]
```diff
@@ -346,6 +507,18 @@
         return retVal;
     };
 
+    /**
+     * Returns and object reprsentation of row data of a widget. This function is normally used to return an object containing the data about what was clicked on a chart/widget.
```
Comment (dtsnell4) [inline src/common/Widget.js:511]: "Returns and" -> "Returns an"
"reprsentation" -> "representation"

File: src/common/Widget.js [hunk 4]
```diff
@@ -346,6 +507,18 @@
...
+     * @method rowToObj
+     * @memberof common_Widget
+     * @instance
+     * @param [Mixed[]] row A row of data from .data() of a widget
```
Comment (dtsnell4) [inline src/common/Widget.js:515]: remove first occurrence of "row".  Also shouldn't this be preceded by {Object}?

File: src/common/Widget.js [hunk 5]
```diff
@@ -346,6 +507,18 @@
...
+     * @returns {Widget}
+     * @example widget
+     * .columns(["ID", "Year 1", "Year 2"])
+     * .data([ [40, 66, 60], [30, 98, 92]  ])
+     * .render();
+     */
```
Comment (dtsnell4) [inline src/common/Widget.js:521]: The example doesn't actually use rowToObj()

File: src/common/Widget.js [hunk 6]
```diff
@@ -649,11 +964,65 @@
...
+     * @example <caption>Example usage of debounce from dermatology.html .</caption>
+     * propEditor.onChange =  Surface.prototype.debounce(function (widget, propID) {
+     *     if (propID === "columns") {
+     *     } else if (propID === "data") {
+     *     } else {
+     *         displaySerialization();
+     *         displaySerializationText();
+     *     }
+     * }, 500);
+     * @see https://github.com/hpcc-systems/Visualization/blob/master/demos/dermatology.html#L143
```
Comment (dtsnell4) [inline src/common/Widget.js:1024]: The line for this example has changed from 143 to 146.  I'm sure its going to keep changing over time.

File: src/common/Widget.js [hunk 7]
```diff
@@ -346,6 +507,18 @@
         return retVal;
     };
 
+    /**
+     * Returns and object reprsentation of row data of a widget. This function is normally used to return an object containing the data about what was clicked on a chart/widget.
+     * @method rowToObj
+     * @memberof common_Widget
+     * @instance
+     * @param [Mixed[]] row A row of data from .data() of a widget
+     * @returns {Widget}
```
Comment (dtsnell4) [inline src/common/Widget.js:516]: Doesn't this return an object? If not need to change that on line 511 "Returns and object reprsentation of row data of a widget.  I'm actually confused about a few of the return types like class(), display(), height(), id()...

File: src/common/Widget.js [hunk 8]
```diff
@@ -649,11 +964,65 @@
...
+     * @memberof common_Widget
+     * @instance
+     * @protected
+     * @param {HTMLElement} domeNode HTML DOMNode of widget container.
+     * @param {D3Selection} element d3 selection object of widget.
+     */
+    Widget.prototype.enter = function (domeNode, element) { };
+
+    /**
+     * The function that is called when this widget "enters" the web page. after enter() and everytime the widget is updated with subsequent render calls.
```
Comment (dtsnell4) [inline src/common/Widget.js:979]: Capitalize "after"

File: src/common/Widget.js [hunk 9]
```diff
@@ -544,6 +809,18 @@
         return null;
     };
 
+    /**
+     * TODO
+     * Returns the absolute position of a widget. { x:[x], y:[y] }
+     * @method getAbsolutePos
+     * @memberof common_Widget
+     * @instance
+     * @param {HTMLElement} domNode.
```
Comment (dtsnell4) [inline src/common/Widget.js:818]: Remove the period after "domNode."

File: src/common/Widget.js [hunk 10]
```diff
@@ -422,6 +643,13 @@
         return this;
     };
 
+    /**
+     * Returns scrollbar width
```
Comment (dtsnell4) [inline src/common/Widget.js:647]: Period at end of line.

File: src/common/Widget.js [hunk 11]
```diff
@@ -281,7 +383,27 @@
...
+     * @return undefined
+     */
+
+    /**
+     * Sets event (click) callback function.
+     * @method on
+     * @memberof common_Widget
+     * @instance
+     * @param {String} eventID The name of the event. Ex: "click".
+     * @param {Widget~EventCb} callback - The callback that is called when event is fired.
```
Comment (dtsnell4) [inline src/common/Widget.js:400]: SHouldn't this be {Function} ?

File: src/api/I2DChart.js [hunk 12]
```diff
@@ -6,11 +11,28 @@
...
+     */
     function I2DChart() {
     }
+    /**
+     * Instance of a HPCC VIZ Palette object.
+     * @member {Object} _palette
+     * @memberof api_I2DChart
+     * @private
+     */
     I2DChart.prototype._palette = Palette.ordinal("default");
```
Comment (dtsnell4) [inline src/api/I2DChart.js:26]: This does not appear on the docs page for I2DChart

File: src/api/ITree.js [hunk 13]
```diff
@@ -6,11 +11,27 @@
...
+     */
     function ITree() {
     }
+    /**
+     * Instance of a HPCC VIZ Palette object.
+     * @member {Object} _palette
+     * @memberof api_ITree
+     * @private
+     */
     ITree.prototype._palette = Palette.ordinal("default");
```
Comment (dtsnell4) [inline src/api/ITree.js:26]: This method does not appear on the docs page

File: src/api/ITree.js [hunk 14]
```diff
@@ -40,7 +61,13 @@
         return this;
     };
 
-    //  Events  ---
+    /**
+     * (event) Overridable click callback function.
+     * @method click
+     * @memberof api_INDChart
```
Comment (dtsnell4) [inline src/api/ITree.js:67]: Should be "api_ITree"

File: src/api/IGraph.js [hunk 15]
```diff
@@ -6,14 +11,31 @@
...
+     * @class api_IGraph
+     */
     function IGraph() {
     }
 
-    //  Events  ---
+    /**
+     * (event) Overridable click callback function for Graph Vertex (Icon)
+     * @method vertex_click
+     * @memberof api_I1DChart
```
Comment (dtsnell4) [inline src/api/IGraph.js:24]: Should be "api_IGraph"

File: src/api/IGraph.js [hunk 16]
```diff
@@ -6,14 +11,31 @@
...
+     * @param {type} column
+     */
     IGraph.prototype.vertex_click = function (d) {
         console.log("Vertex Click: " + d.id());
     };
 
+    /**
+     * (event) Overridable click callback function for Graph Edge (Line/Arrow).
+     * @method edge_click
+     * @memberof api_I1DChart
```
Comment (dtsnell4) [inline src/api/IGraph.js:35]: Should be "api_IGraph"

File: src/api/I1DChart.js [hunk 17]
```diff
@@ -6,18 +11,41 @@
...
+     */
     function I1DChart() {
     }
+    /**
+     * Instance of a HPCC VIZ Palette object.
+     * @member {Object} _palette
+     * @memberof api_I1DChart
+     * @private
+     */
     I1DChart.prototype._palette = Palette.rainbow("default");
```
Comment (dtsnell4) [inline src/api/I1DChart.js:26]: Does not appear on docs page

File: src/api/INDChart.js [hunk 18]
```diff
@@ -6,11 +11,27 @@
...
+     */
     function INDChart() {
     }
+    /**
+     * Instance of a HPCC VIZ Palette object.
+     * @member {Object} _palette
+     * @memberof api_INDChart
+     * @private
+     */
     INDChart.prototype._palette = Palette.ordinal("default");
```
Comment (dtsnell4) [inline src/api/INDChart.js:26]: Does not appear in docs page

File: src/common/SVGWidget.js [hunk 19]
```diff
@@ -69,6 +126,16 @@
         return retVal;
     };
 
+    /**
+     * Sets target container for widget to render in. (Normally a DIV)
+     * @method target
+     * @memberof common_SVGWidget
+     * @instance
+     * @param {String|HTMLElement} [size] An object with the properties "width" and "height".
```
Comment (dtsnell4) [inline src/common/SVGWidget.js:134]: Wrong param

File: src/form/ISlider.js [hunk 20]
```diff
@@ -6,18 +11,51 @@
...
+     * @memberof form_ISlider
+     * @param {type} value
+     */
     ISlider.prototype.click = function (value) {
         console.log("click:  " + value);
     };
+    /**
+     * Slider selection event callback function.
+     * @method click
+     * @memberof common_IList
```
Comment (dtsnell4) [inline src/form/ISlider.js:55]: IList should be Slider

File: src/common/IMenu.js [hunk 21]
```diff
@@ -6,25 +11,52 @@
...
+     * Overridable function that is fired before menu is shown.
+     * @method preShowMenu
+     * @memberof common_IMenu
+     * @param {type} d
+     */
+    IMenu.prototype.preShowMenu = function (d) {
         console.log("preShowMenu");
     };
+    /**
+     * Overridable function that is fired before menu is shown.
```
Comment (dtsnell4) [inline src/common/IMenu.js:55]: Should be "after menu is hidden".

File: src/api/I2DChart.js (hunk 12 above)
Comment (mzummo) [inline src/api/I2DChart.js:26]: have to pass -p flag when building docs

File: src/api/ITree.js (hunk 13 above)
Comment (mzummo) [inline src/api/ITree.js:26]: have to pass -p flag when building docs

File: src/api/I1DChart.js (hunk 17 above)
Comment (mzummo) [inline src/api/I1DChart.js:26]: need -p flag

File: src/api/INDChart.js (hunk 18 above)
Comment (mzummo) [inline src/api/INDChart.js:26]: need -p flag

**Pull request closed without merging**